src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+
+use crate::TransformArgs;
+use expr::{AggregateExpr, AggregateFunc, MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer key is not a subset of inner key
+                if !group_key
+                    .iter()
+                    .all(|e| matches!(e, MirScalarExpr::Column(_)) && inner_group_key.contains(e))
+                {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
+                    // Replace inner reduce with map + project (no grouping)
+                    let mut outputs = vec![];
+                    let mut scalars = vec![];
+                    for e in inner_group_key {
+                        if let MirScalarExpr::Column(i) = e {
+                            outputs.push(*i);
+                        } else {
+                            scalars.push(e.clone());
```

Comment:
This alters the order in which the columns are projected by the resulting subplan without the reduction and, in order for the resulting plan to be correct, the order of the projected columns must be preserved. To preserve the order, you just need to remove the loop below and add the line within the loop in this else branch as:

```
outputs.push(arity + scalars.len());
scalars.push(e.clone());
```

IINM with the current code the following example:

```
materialize=> create table a(a int, b int, c int);
CREATE TABLE
materialize=> explain select a from (select a from a group by b / 10, a) a group by a;
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.a (u21)+
 | Distinct group=((#1 / 10), #0)+
 | Distinct group=(#1)           +
 
(1 row)

```
will result in:

```
materialize=> explain select a from (select a from a group by b / 10, a) a group by a;
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.a (u21)+
 | Map (#1 / 10)    +
 | Project (#0, #3)
 | Distinct group=(#1)           + // grouping by a/10
 
(1 row)

materialize=> 
```

rather than:

```
materialize=> explain select a from (select a from a group by b / 10, a) a group by a;
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.a (u21)+
 | Map (#1 / 10)    +
 | Project (#3, #1)
 | Distinct group=(#1)           + // grouping by a
 
(1 row)

materialize=> 

```

Reply:
Good catch. I hadn't thought of this. Updated.